cachi2/core/package_managers/yarn_classic/workspaces.py

```diff
@@ -0,0 +1,114 @@
+import json
+from contextlib import suppress
+from itertools import chain
+from pathlib import Path
+from typing import Any, Iterable
+
+import pydantic
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.models.input import YarnClassicPackageInput
+from cachi2.core.rooted_path import RootedPath
+
+
+class Workspace(pydantic.BaseModel):
+    """Workspace model."""
+
+    path: Path  # path to a workspace.
+    package_contents: dict  # package data extracted from path/"package.json".
+    # package reference for potential nested workspace extraction:
+    package: YarnClassicPackageInput
+
+
+def ensure_no_path_leads_out(
+    paths: Iterable[Path],
+    source_dir: RootedPath,
+) -> None:
+    """Ensure no path leads out of source directory.
+
+    Raises an exception when any path is not relative to source directory.
+    Does nothing when path does not exist in the file system.
+    """
+    for path in paths:
+        if not path.is_relative_to(source_dir.path):
+            raise PackageRejected(
+                f"Found a workspace path which is not relative to package: {path}",
+                solution=(
+                    "Avoid using packages which try to access your filesystem "
+                    "outside of package directory."
+                ),
+            )
+
+
+def ensure_workspaces_are_well_formed(
+    paths: Iterable[Path],
+) -> None:
+    """Ensure that every workspace contains package.json.
+
+    Reject the package otherwise.
+    """
+    for p in paths:
+        if not Path(p, "package.json").is_file():
+            raise PackageRejected(
+                reason=f"Workspace {p} does not contain 'package.json'",
+                solution=None,
+            )
```

Comment:
> Checking that some file is present somewhere is not models responsibility. Someone else must do this

I'd argue that the model is to represent a workspace which *is defined** by the presence of a `package.json` file; no `package.json` file -> no package or workspace in this case.

Fair enough though, given the not-the-most-pleasant pydantic model interface when it comes to testing the validation, it's not a blocker for me -> resolving.

Reply:
> A model represents some structured data, its responsibility is to make sure that all necessary fields were supplied.

Actually, since you're missing a validator for the `path` attribute, you might as well extend the model with a `model_validator(after)` one that checks `package.json` presence as it's closely related.
I have found 
- https://github.com/containerbuildsystem/cachi2/blob/c94a1a24c1777036bbed121281c41be1b8f156f1/cachi2/core/models/input.py#L70
- https://github.com/containerbuildsystem/cachi2/blob/c94a1a24c1777036bbed121281c41be1b8f156f1/cachi2/core/models/input.py#L154
- [most recent one] https://github.com/containerbuildsystem/cachi2/blob/c94a1a24c1777036bbed121281c41be1b8f156f1/cachi2/core/models/input.py#L88

which are similar but definitely not trying to do a deep file system checking. So, either we've already broken good data-logic separation there as well or we could do something similar and a bit more here.

Reply:
I am sorry, but mode="after" is useless: package.json is used to instantiate a Workspace, moreover, there is a validator that makes sure that package.json contains a name. By the time we get to "after" we have already proven that all paths are correct. 'path' field is validated to be a correct Path-like string though, and we also check that no path leads out of the model by checking that it is joinable within the root. Moreover, testing any filesystem layout is still not a job for a model, it becomes bound to a specific filesystem otherwise. It is already a little mixed up by having package.json contents in it, I don't think it should be bound to a FS tighter than that.